### PR TITLE
Array Assignment Amarja Q1 TO Q11

### PR DESCRIPTION
**Pull request by AmarjaBTechnocredits**
Please review

### CODE REVIEW

File: JavaAssignments/src/amarja/AllDiffindex.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.Scanner;
+//Question 9 Program to find all different index from given array
+public class AllDiffindex {
+	public static void main(String args[])
+	{		Scanner sc=new Scanner(System.in);//Reading the size and values of array 1
+			System.out.println("Enter size of first array:");
+			int size1=sc.nextInt();
+			System.out.println("Enter values of first array:");
+			int[]num1 = new int[size1];
+			for (int index=0;index<size1;index++)
```
Comment (mkanani13) [inline JavaAssignments/src/amarja/AllDiffindex.java]: Amarja, here you wrote same logic two times to get data from user. Tomorrow if I asked to find differ index from 5 arrays then same logic you need to write for 5 times. always better to create function which will return int array of user given size.

File: JavaAssignments/src/amarja/AllDiffindex.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+			}
+			sc.close();
+			AllDiffindex diff=new AllDiffindex();
+			diff.findeqarrind(num1,num2);	//Function to find all different index
+	}
+	void findeqarrind(int num1[],int num2[])
+	{					
+			for (int index=0;index<num1.length;index++)
+				{
+					if (num1[index]!= num2[index])
```
Comment (mkanani13) [inline JavaAssignments/src/amarja/AllDiffindex.java]: We should consider negative scenarios also while writing code. for example what if num1 having more value then num2 . here you are iterating loop till num1.length , if num2 has less elements then comparison num1[index] != num2[index] will throw arrayindex out of bound exception.

File: JavaAssignments/src/amarja/Arraycontnum.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package amarja;
+import java.util.*;
+//Question 3 Program to find array contains number 
+public class Arraycontnum {
+	boolean findnum(int arr[],int num)
+	{	boolean flag=false;
+		for (int index=0;index<arr.length;index++)
+		{
+			if (arr[index]==num)
```
Comment (mkanani13) [inline JavaAssignments/src/amarja/Arraycontnum.java]: Once you find the number, you should break this loop.
Consider a scenario where there are 100 elements and you find first element as your number then also loop will iterate 99 time extra which is not required.

File: JavaAssignments/src/amarja/Arraycontnum.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+				flag=true;
+		}
+		return flag;
+	}
+	public static void main (String args[])
+	{
+		Scanner sc=new Scanner(System.in);
+		System.out.println("Enter size of array");
+		int size=sc.nextInt();
+		System.out.println("Enter values of array");
```
Comment (mkanani13) [inline JavaAssignments/src/amarja/Arraycontnum.java]: Please create another function to take int value from user. 
Write minimum code in main method as main method logic wont be reusable by other class.

File: JavaAssignments/src/amarja/Equalarrays.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+		if (num1.length != num2.length)
+			return flag;
+		else 
+			{
+				for (int index=0;index<num1.length;index++)
+				{
+					if (num1[index]== num2[index])
+							flag=true;
+					else
+							flag=false;
```
Comment (mkanani13) [inline JavaAssignments/src/amarja/Equalarrays.java:17]: break statement required here.

File: JavaAssignments/src/amarja/Findmaxminnum.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package amarja;
+//Question 4 Program to find the maximum and minimum number in array
+public class Findmaxminnum {
+		int maximum(int num [])
+		{
+			int max=num[0];
+			for (int i=0;i<num.length;i++)
```
Comment (mkanani13) [inline JavaAssignments/src/amarja/Findmaxminnum.java]: you should start loop from i=1 as max is defined as num[0]

File: JavaAssignments/src/amarja/Firstdiffindex.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.Scanner;
+//Question 8 Program to find first different index from given array
+public class Firstdiffindex {
+		public static void main(String args[])
+		{		Scanner sc=new Scanner(System.in);//Reading the size and values of array 1
+				System.out.println("Enter size of first array:");
+				int size1=sc.nextInt();
+				System.out.println("Enter values of first array:");
+				int[]num1 = new int[size1];
+				for (int index=0;index<size1;index++)
```
Comment (mkanani13) [inline JavaAssignments/src/amarja/Firstdiffindex.java]: As discussed above , write separate function to take data from user.

File: JavaAssignments/src/amarja/Firstdiffindex.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+				int[]num2 = new int[size2];
+				for (int index=0;index<size2;index++)
+				{
+					num2[index]=sc.nextInt();
+				}
+				sc.close();
+				Firstdiffindex diff=new Firstdiffindex();
+				diff.finddiffarrind(num1,num2);
+		}
+		void finddiffarrind(int num1[],int num2[])
```
Comment (mkanani13) [inline JavaAssignments/src/amarja/Firstdiffindex.java]: This logic can be written in better way.

boolean finddiffarrind(int num1[],int num2[])
{				
             for(int index=0;index<num1.length;index++)
	     {
			if (num1[index]!= num2[index])
			{	
					System.out.println("Values are not matching at index " +index);	
					return false;	
			}				
	      }
              return true;
}

Then you should check return value, if value is true then both arrays must be equal.

File: JavaAssignments/src/amarja/Secondmax.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+				max=num[i];
+			}
+		}
+		return secmax;	
+		}
+	public static void main(String[] args) {
+		Secondmax smax=new Secondmax();
+		Takeinputarrint input = new Takeinputarrint();//Accepting array input using method
+		int num[]=input.takeintegerinputuser();
+		System.out.println("Second maximum number is  "+smax.secmaximum(num));
```
Comment (mkanani13) [inline JavaAssignments/src/amarja/Secondmax.java:26]: Your logic wont work for below input.
int num[]={10,14,99,110,108};

please check !

File: JavaAssignments/src/amarja/Sumequaltonum.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
+package amarja;
+//Question 7 Program to find all pair of integer array whose sum is equal to given number
+import java.util.Scanner;
+public class Sumequaltonum {
+	void findpair(int arr[],int no)
```
Comment (mkanani13) [inline JavaAssignments/src/amarja/Sumequaltonum.java:5]: We need to discuss this logic.
Please contact me whenever possible.

File: JavaAssignments/src/amarja/Treecount.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+		 }
+		 return count;
+	 }
+	// Main Function calling another class to take integer array input 
+	public static void main(String[] args)
+		{	
+		Takeinputarrint input = new Takeinputarrint();
+		int num[]=input.takeintegerinputuser();
+		Treecount treecount= new Treecount();
+		int count=treecount.findtreecount(num);
```
Comment (mkanani13) [inline JavaAssignments/src/amarja/Treecount.java:22]: Your program produces wrong output for below input.

int num[]={3,9,2,14,3,13};

Please try to check your program logic with multiple inputs.

**Review by mkanani13 — CHANGES_REQUESTED** [review]
Amarja, Please change your code as per review comments and push it back on same branch.

File: JavaAssignments/src/amarja/Arraycontnum.java (hunk 4 above)
Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/Arraycontnum.java]: I had created another function and was using it eveywhere but I was getting the connection error for input stream.Hence had to write it in main method
Exception in thread "main" java.lang.NullPointerException
	at amarja.Arraycontnum.main(Arraycontnum.java:22)


please sugest ..As per your suggestion on skyp ehad written in main method.still getting errir

Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/Arraycontnum.java]: Tried again able to create a seperate function .Comments incorporate.Please review

File: JavaAssignments/src/amarja/Equalarrays.java (hunk 5 above)
Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/Equalarrays.java:17]: Incorporated comments .Break statement added and seperate function created for input

File: JavaAssignments/src/amarja/Findmaxminnum.java (hunk 6 above)
Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/Findmaxminnum.java]: Incorporated review comments as sugested

File: JavaAssignments/src/amarja/AllDiffindex.java (hunk 1 above)
Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/AllDiffindex.java]: Incorporated review comments as suggested

File: JavaAssignments/src/amarja/Firstdiffindex.java (hunk 7 above)
Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/Firstdiffindex.java]: Incorporated review comments as suggested

File: JavaAssignments/src/amarja/Firstdiffindex.java (hunk 8 above)
Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/Firstdiffindex.java]: Incorporated better logic as suggested

File: JavaAssignments/src/amarja/Treecount.java (hunk 11 above)
Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/Treecount.java:22]: Modified the program

Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/Treecount.java:22]: tested with multiple scenarios

File: JavaAssignments/src/amarja/Sumequaltonum.java (hunk 10 above)
Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/Sumequaltonum.java:5]: ok we will disucss on saturday

File: JavaAssignments/src/amarja/Secondmax.java (hunk 9 above)
Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/Secondmax.java:26]: Changed the logic and tested for other combinations as well

File: JavaAssignments/src/amarja/AllDiffindex.java (hunk 2 above)
Comment (AmarjaBTechnocredits) [reply JavaAssignments/src/amarja/AllDiffindex.java]: Amended the code to consider if the array lengths are not equal repeat the loop till the minimum length

**Review by mkanani13 — APPROVED** [review]
Code looks good to me.

**Comment by AmarjaBTechnocredits** [comment]
Resolved conflicts

**Review by mkanani13 — APPROVED** [review]

**Pull request merged**